nupic/torch/functions/k_winners.py

```diff
@@ -21,229 +21,187 @@
-          The activity of the top k units will be allowed to remain, the rest are
-          set to zero.
-
-        :param boost_strength:
-          A boost strength of 0.0 has no effect on x.
-
-        :return:
-          A tensor representing the activity of x after k-winner take all.
-        """
-
-        # Take the boosted version of the input x, find the top k winners.
-        # Compute an output that only contains the values of x corresponding to
-        # the top k boosted values. The rest of the elements in the output
-        # should be 0.
-        batch_size = x.shape[0]
-        boosted = boost_activations(x, duty_cycles, boost_strength)
-        boosted = boosted.reshape((batch_size, -1))
-        xr = x.reshape((batch_size, -1))
-        res = torch.zeros_like(boosted)
-        topk, indices = boosted.topk(k, dim=1, sorted=False)
-        res.scatter_(1, indices, xr.gather(1, indices))
-        res = res.reshape(x.shape)
-
-        ctx.save_for_backward(indices)
-        return res
-
-    @staticmethod
-    def backward(ctx, grad_output):
-        """In the backward pass, we set the gradient to 1 for the winning
-        units, and 0 for the others.
-        """
-        batch_size = grad_output.shape[0]
-        indices, = ctx.saved_tensors
-
-        g = grad_output.reshape((batch_size, -1))
-        grad_x = torch.zeros_like(g, requires_grad=False)
-        grad_x.scatter_(1, indices, g.gather(1, indices))
-        grad_x = grad_x.reshape(grad_output.shape)
-
-        return grad_x, None, None, None
-
-
-class KWinners2dLocal(torch.autograd.Function):
+    :param relu:
+      Whether to simulate the effect of applying ReLU before KWinners
+
+    :param inplace:
+      Whether to modify x in place
+
+    :return:
+      A tensor representing the activity of x after k-winner take all.
     """
-    A K-winner take all autograd function for CNN 2D inputs (batch, Channel, H,
-    W) where k-winners are chosen independently for each location. Local k-winners
-    selects the top k channels locally for each of the H X W locations.
+    if k == 0:
+        return torch.zeros_like(x)
+
+    boosted = boost_activations(x, duty_cycles, boost_strength)
+
+    if break_ties:
+        indices = boosted.topk(k=k, dim=1, sorted=False)[1]
+        off_mask = torch.ones_like(boosted, dtype=torch.bool)
+        off_mask.scatter_(1, indices, False)
+
+        if relu:
+            off_mask |= (boosted <= 0)
+    else:
+        threshold = boosted.kthvalue(x.shape[1] - k + 1, dim=1,
+                                     keepdim=True)[0]
+
+        if relu:
+            threshold.clamp_(min=0)
+        off_mask = boosted < threshold
+
+    if inplace:
+        return x.masked_fill_(off_mask, 0)
+    else:
+        return x.masked_fill(off_mask, 0)
+
 
-    .. seealso::      Function :class:`KWinners2dGlobal`
+def kwinners2d(x, duty_cycles, k, boost_strength, local=True, break_ties=True,
+               relu=False, inplace=False):
```

Comment:
Maybe we should default to break_ties=False assuming it is unlikely it will impact the accuracy on most common use cases (float32)